### ISSUE TITLE
Share parameters between LSTMs

### ISSUE THREAD

**Issue by nmduc** [Element-Research/rnn#215]
What is the correct way to share parameters between two LSTMs.
There is 'shared' function for sharing parameters between nn.Modules but I'm not sure whether using it on LSTM will work as expected or not?

**Comment by nmduc** [comment]
I am trying with the following architecture, but always got zero gradients so cannot train the model at all. This problem remains even if I don't share parameters between two LSTMs.

I have tried substitute the two recurrent modules with nn.Linear modules and the gradient could still flow well to the Linear modules. So there seems to be a problem with the Sequencer here. 

`
require 'torch'
require 'nn'
require 'nngraph'
require 'rnn'
dtype = 'torch.CudaTensor'

local data_dim = 300
local hidden_dim = 256
local rho = 10

local lstm1 = nn.LSTM(data_dim, hidden_dim)
local lstm2 = nn.LSTM(data_dim, hidden_dim)
lstm1.recurrentModule:share(lstm2.recurrentModule, 'weight')
lstm1.recurrentModule:share(lstm2.recurrentModule, 'bias')

recurrent1 = nn.Sequential()
            :add(nn.SplitTable(2))
            :add(nn.Sequencer(lstm1))
            :add(nn.SelectTable(-1))
            
recurrent2 = nn.Sequential()
            :add(nn.SplitTable(2))
            :add(nn.Sequencer(lstm2))
            :add(nn.SelectTable(-1))

local parallel1 = nn.ParallelTable():add(recurrent1):add(recurrent2)

-- input must be of shape: batch_size x length x data_dim x 2 (2 for a pair of sequence at a time)
local net = nn.Sequential()
            :add(nn.SplitTable(4))
            :add(parallel1)
            :add(nn.PairwiseDistance(1))   
            :add(nn.MulConstant(-1))       -
            :add(nn.Exp())

return net
`

**Comment by nicholas-leonard** [comment]
@nmduc Using [sharedClone](https://github.com/Element-Research/dpnn#clone-modulesharedcloneshareparams-sharegradparams) should be easier : 
```lua
lstm1 = nn.FastLSTM(inputsize, outputsize)
lstm2 = lstm1:sharedClone()
```

**Comment by nmduc** [comment]
@nicholas-leonard Thanks for pointing it out. It looks much cleaner.
So with sharedClone, by default, both params and gradParams will be shared by two modules. In this case, how is gradParams calculated? Is it the sum of gradParams from both modules or is it just equal gradParams of the original modules?

Btw, my last problem with zero gradients was caused by buggy SelectTable in my nn installation. After re-installing it, gradients flow well.
